mithril-aggregator/src/database/provider/open_message.rs

```diff
@@ -0,0 +1,481 @@
+        let parameters = vec![
+            Value::String(Uuid::new_v4().to_string()),
+            Value::Integer(epoch.0 as i64),
+            Value::String(serde_json::to_string(beacon)?),
+            Value::Integer(*signed_entity_type as i64),
+            Value::String(message.to_string()),
+        ];
+
+        Ok(WhereCondition::new(expression, parameters))
+    }
+}
+
+impl<'client> Provider<'client> for InsertOpenMessageProvider<'client> {
+    type Entity = OpenMessage;
+
+    fn get_connection(&'client self) -> &'client Connection {
+        self.connection
+    }
+
+    fn get_definition(&self, condition: &str) -> String {
+        let aliases = SourceAlias::new(&[("{:open_message:}", "open_message")]);
+        let projection = Self::Entity::get_projection().expand(aliases);
+
+        format!("insert into open_message {condition} returning {projection}")
+    }
+}
+
+struct DeleteOpenMessageProvider<'client> {
+    connection: &'client Connection,
+}
+
+impl<'client> DeleteOpenMessageProvider<'client> {
+    pub fn new(connection: &'client Connection) -> Self {
+        Self { connection }
+    }
+
+    fn get_epoch_condition(&self, epoch: Epoch) -> WhereCondition {
+        WhereCondition::new(
+            "epoch_setting_id = ?*",
+            vec![Value::Integer(epoch.0 as i64)],
+        )
+    }
+}
+
+impl<'client> Provider<'client> for DeleteOpenMessageProvider<'client> {
+    type Entity = OpenMessage;
+
+    fn get_connection(&'client self) -> &'client Connection {
+        self.connection
+    }
+
+    fn get_definition(&self, condition: &str) -> String {
+        let aliases = SourceAlias::new(&[("{:open_message:}", "open_message")]);
+        let projection = Self::Entity::get_projection().expand(aliases);
+
+        format!("delete from open_message where {condition} returning {projection}")
+    }
+}
+
+/// ## Open message repository
+///
+/// This is a business oriented layer to perform actions on the database through
+/// providers.
+pub struct OpenMessageRepository {
+    connection: Arc<Mutex<Connection>>,
+}
+
+impl OpenMessageRepository {
+    /// Instanciate service
+    pub fn new(connection: Arc<Mutex<Connection>>) -> Self {
+        Self { connection }
+    }
+
+    /// Return the latest [OpenMessage] for the given Epoch and [SignedEntityType].
+    pub async fn get_open_message(
+        &self,
+        epoch: Epoch,
+        signed_entity_type: &SignedEntityType,
+    ) -> StdResult<Option<OpenMessage>> {
+        let lock = self.connection.lock().await;
+        let provider = OpenMessageProvider::new(&lock);
+        let filters = provider
+            .get_epoch_condition(epoch)
+            .and_where(provider.get_signed_entity_type_condition(signed_entity_type));
+        let mut messages = provider.find(filters)?;
+
+        Ok(messages.next())
+    }
+
+    /// Create a new [OpenMessage] in the database.
+    pub async fn create_open_message(
```

Comment:
If there is already an open message and a newer one is created for the same epoch, we should remove the previous one? WDYT?